weasyl/login.py

```diff
@@ -173,37 +173,62 @@ def create(form):
     if sysname in ["admin", "administrator", "mod", "moderator", "weasyl",
                    "weasyladmin", "weasylmod", "staff", "security"]:
         raise WeasylError("usernameInvalid")
-    if email_exists(email):
-        raise WeasylError("emailExists")
     if username_exists(sysname):
         raise WeasylError("usernameExists")
 
-    # Create pending account
+    # Account verification token
     token = security.generate_key(40)
 
-    d.engine.execute(d.meta.tables["logincreate"].insert(), {
-        "token": token,
-        "username": username,
-        "login_name": sysname,
-        "hashpass": passhash(password),
-        "email": email,
-        "birthday": birthday,
-        "unixtime": arrow.now(),
-    })
+    # Only attempt to create the account if the email is unused (as defined by the function)
+    if not email_exists(email):
+        # Create pending account
+        d.engine.execute(d.meta.tables["logincreate"].insert(), {
+            "token": token,
+            "username": username,
+            "login_name": sysname,
+            "hashpass": passhash(password),
+            "email": email,
+            "birthday": birthday,
+            "unixtime": arrow.now(),
+        })
 
-    # Queue verification email
-    emailer.append([email], None, "Weasyl Account Creation", d.render(
-        "email/verify_account.html", [token, sysname]))
-    d.metric('increment', 'createdusers')
+        # Queue verification email
+        emailer.append([email], None, "Weasyl Account Creation", d.render(
+            "email/verify_account.html", [token, sysname]))
+        d.metric('increment', 'createdusers')
+    else:
+        # Store a dummy record to support plausible deniability of email addresses
+        # So "reserve" the username, but mark the record invalid, and use the token to satisfy the uniqueness
+        #  constraint for the email field (e.g., if there is already a valid, pending row in the table).
+        d.engine.execute(d.meta.tables["logincreate"].insert(), {
+            "token": token,
+            "username": username,
+            "login_name": sysname,
+            "hashpass": '',
```

Comment:
It’s probably not practical to eliminate all timing attacks the way the tables are set up right now, but this one seems particularly easy to measure – bcrypt with our current work factor of 13 takes a few hundred milliseconds. If an attacker times the request, they can determine to some certainty whether the e-mail is registered. I think `passhash(password)` should be run here too to avoid that.